onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,441 @@
 
 #include "core/framework/data_types.h"
 #include "core/framework/sparse_tensor.h"
+#include "core/framework/data_transfer_manager.h"
+#include "core/framework/utils.h"
+
+#include <safeint/SafeInt.hpp>
 
 using namespace onnxruntime::common;
 
 namespace onnxruntime {
 
-SparseTensor::SparseTensor(MLDataType elt_type,
-                           const TensorShape& shape,
-                           size_t nnz,
-                           void* values_data,
-                           void* indices_data,
-                           const OrtMemoryInfo& memory_info)
-    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), values_data, memory_info),
-      indices_(DataTypeImpl::GetType<int64_t>(),
-               TensorShape({static_cast<int64_t>(nnz), static_cast<int64_t>(shape.NumDimensions())}),
-               indices_data, memory_info, 0),
-      shape_(shape) {}
-
-SparseTensor::SparseTensor(MLDataType elt_type,
-                           const TensorShape& shape,
-                           size_t nnz,
+std::ostream& operator<<(std::ostream& os, SparseFormat flags) {
+  return os << std::hex << static_cast<uint32_t>(flags);
+}
+
+namespace {
+// Round up size to a multiple of align.
+// Example:
+//   Roundup(13, 5)   => 15
```

Comment:
these examples still use the two argument version of Roundup
"multiple of align" should be updated too